### ISSUE TITLE
Batch norm slows down custom convolution

### ISSUE THREAD

**Issue by tom-bird** [google/jax#1150]
Hi,

I noticed that running a model with a custom implementation of a convolution resulted in massive slowdown when batch norm was used. When batch norm is not used, the model runs in similar time to using the stax convolution.

I have included some code below to reproduce. The code runs a custom version of a 1d conv using `lax.gather` and then `np.dot`, and compares the running time to using `stax.GeneralConv`.

Running the code with `bn=False` (i.e. no batch norm) I get:
```
Normal conv: 0.026s
My conv: 0.023s
```
and then with `bn=True` I get:
```
Normal conv: 0.029s
My conv: 0.609s
```
So the run time with batch norm is about 25x larger with my custom convolution, whereas it doesnt really affect the stax conv. The difference is slightly smaller but still large without jit compiling.

I am running 
```
jax==0.1.41
jaxlib==0.1.23
```
on GPU, pip installed as per the jax homepage, with CUDA 9.0, CUDNN 7.6.2.

Here is the code. Note that the convolution is not quite right - if you change the in channels to be > 1 then it will not match (not reshaping the filter and params correctly before the `np.dot`). But for 1 input channel it is correct - this is checked in the below script.

```
import functools

import jax.numpy as np
import numpy as onp
from jax import jit, random, lax
from jax.experimental import stax
import timeit

bn = True

batch_size = 5
x_len = 10000
in_chan = 1
out_chan = 2
filter_len = 3
use_jit = True

a = onp.random.randn(batch_size, x_len, in_chan)

Conv1D = functools.partial(stax.GeneralConv, ('NHC', 'HIO', 'NHC'))

def apply_1d_conv(params, input, **kwargs):
    """perform a 1d conv manually"""
    W, b = params
    # get matrix of slices
    start_indices = onp.array([(0, i, 0) for i in range(input.shape[1] - filter_len + 1)])
    dimension_numbers = lax.GatherDimensionNumbers(
        offset_dims=(0, 1, 2),
        collapsed_slice_dims=(),
        start_index_map=(0, 1, 2)
    )
    col = lax.gather(input,
                     start_indices=start_indices,
                     dimension_numbers=dimension_numbers,
                     slice_sizes=(batch_size, filter_len, in_chan))
    col = np.transpose(col, (0, 3, 1, 2))
    col = np.reshape(col, (col.shape[0], col.shape[1], -1), order='F')
    return np.dot(col, np.reshape(W, (-1, out_chan), order='F')) + b

def myConv1D(out_chans, filter_shape):
    init_fun = Conv1D(out_chans, filter_shape)[0]
    return init_fun, apply_1d_conv

model = stax.serial(Conv1D(out_chan, (filter_len,)), stax.BatchNorm(axis=(0, 1)) if bn else stax.Identity)
model2 = stax.serial(myConv1D(out_chan, (filter_len,)), stax.BatchNorm(axis=(0, 1)) if bn else stax.Identity)

init_fun, apply_fun = model
_, apply_fun2 = model2
rng = random.PRNGKey(0)
params = init_fun(rng, a.shape)[1]

f = lambda: jit(apply_fun)(params, a).block_until_ready() if use_jit else apply_fun(params, a).block_until_ready()
g = lambda: jit(apply_fun2)(params, a).block_until_ready() if use_jit else apply_fun2(params, a).block_until_ready()

out1 = apply_fun(params, a)
out2 = apply_fun2(params, a)
onp.testing.assert_allclose(out1, out2)  # check the convs are the same

f(), g()  # run once to compile stuff
t0, t1 = timeit.timeit(f, number=100), timeit.timeit(g, number=100)
print('Normal conv: {:.3f}s'.format(t0))
print('My conv: {:.3f}s'.format(t1))
```
Any ideas as to what could be happening?

**Comment by hawkinsp** [comment]
Thanks for the report! I raised this as an issue with the XLA team. My guess is that XLA is generating a bad fusion of the two computations.

(You can probably work around by introducing a `jit` boundary between the two layers, although that might have performance problems of its own.)

**Comment by mattjj** [comment]
Thanks to Peter raising this with the XLA folks, we learned that this is meant to be fixed in tensorflow/tensorflow#31259. To get this fix into JAX we need to (1) update the XLA pointer in our build scripts ([here's an example commit of how that looks](https://github.com/google/jax/commit/92a33b8b253b299bf95e55048e4dc1c1551d5c6c)) and (2) build and upload the jaxlib wheels. After the first step you can also build it yourself.

**Comment by hawkinsp** [comment]
PR #1199 merged the fix. It requires a Jaxlib rebuild, so you'll either have to build jaxlib yourself or wait until we push new wheels.

**Comment by tom-bird** [comment]
Thanks for your help!
